compiler/optimizer/VirtualGuardHeadMerger.cpp

```diff
@@ -260,28 +292,41 @@ int32_t TR_VirtualGuardHeadMerger::perform() {
                if (trace())
                   traceMsg(comp(), "  Created new block_%d to hold guard [%p] from block_%d\n", guard2Block->getNumber(), guard2, guard2Block->getNumber());
 
-               // the block created above guard2 contains only privarg treetops if
-               // guard2 is a runtime-patchable guard. We need to move these args up
-               // to the insert point and update the insertion pointers as needed
-
-               //TODO We can safely leave code ahead of an HCR guard in place if no runtime guards are going to be moved or if
-               //     there is no data dependence between the code moved from ahead of the runtime guards and the code ahead
-               //     of the HCR guard. This opt will increase the chances of a patch point not requiring nops (perf opportunity).
-			   TR::Block *privargBlock = guard2Block->getPrevBlock();
-
-			   if (trace())
-				  traceMsg(comp(), "  Moving privarg block_%d after block_%d\n", privargBlock->getNumber(), privargIns->getNumber());
-
-			   moveBlockAfterDest(cfg, privargBlock, privargIns);
-
-			   if (HCRIns == privargIns)
-				  HCRIns = privargBlock;
-			   if (runtimeIns == privargIns)
-				  runtimeIns = privargBlock;
-			   privargIns = privargBlock;
-
-			   // refresh the insertPoint since it could be stale after the above updates
-			   insertPoint = runtimeIns;
+               // We should leave code ahead of an HCR guard in place because:
+               // 1, it might have side effect to runtime guards after it, moving it up might cause us to falsely merge
+               //    the subsequent runtime guards
+               // 2, it might contain live monitor, moving it up above a guard can affect the monitor's live range
+               if (!guard2->isHCRGuard())
```

Comment:
As per discuss with Andrew, we need a extra check to decide whether it's safe to move the runtime guard, i.e. if there are priv arg stores for other calls ahead of the runtime guard, it's not safe to move it. Builds are running and will remove WIP once they have passed.